src/rust-crypto/PerSessionKeyBackupDownloader.ts

```diff
@@ -0,0 +1,469 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import * as RustSdkCryptoJs from "@matrix-org/matrix-sdk-crypto-wasm";
+import { OlmMachine } from "@matrix-org/matrix-sdk-crypto-wasm";
+
+import { Curve25519AuthData, KeyBackupSession } from "../crypto-api/keybackup";
+import { Logger } from "../logger";
+import { ClientPrefix, IHttpOpts, MatrixError, MatrixHttpApi, Method } from "../http-api";
+import { RustBackupManager } from "./backup";
+import { CryptoEvent } from "../matrix";
+import { encodeUri, sleep } from "../utils";
+import { BackupDecryptor } from "../common-crypto/CryptoBackend";
+
+/**
+ * Enumerates the different kind of errors that can occurs when downloading and importing a key from backup.
+ */
+enum KeyDownloadError {
+    /** The backup version in use is out of sync with the server version. */
+    VERSION_MISMATCH = "VERSION_MISMATCH",
+    /** The requested key is not in the backup. */
+    MISSING_DECRYPTION_KEY = "MISSING_DECRYPTION_KEY",
+    /** A network error occurred while trying to get the key. */
+    NETWORK_ERROR = "NETWORK_ERROR",
+    /** The loop as been stopped. */
+    STOPPED = "STOPPED",
+    /** An unknown error occurred while decrypting/importing the key */
+    UNKNOWN_ERROR = "UNKNOWN_ERROR",
+    /** The server is rate limiting us. */
+    RATE_LIMITED = "RATE_LIMITED",
+    /** The backup is not configured correctly, can be that there is no backup, that it's not trusted
+     * , that we don't have the correct key in cache... */
+    CONFIGURATION_ERROR = "CONFIGURATION_ERROR",
+}
+
+/** Helper type for requested session*/
+type SessionInfo = { roomId: string; megolmSessionId: string };
+
+/** Helper type for the result of a key download. */
+type KeyDownloadResult =
+    | { ok: true; value: KeyBackupSession }
+    | { ok: false; error: KeyDownloadError; [key: string]: any };
```

Comment:
the idiomatic way to do this would be to have `queryKeyBackup` throw a [custom error type](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#custom_error_types):


```
enum KeyDownloadErrorCode {
    VERSION_MISMATCH = "VERSION_MISMATCH",
    // etc
}

class KeyDownloadError extends Error {
    constructor(private readonly code: KeyDownloadErrorCode) {
        super(code);
    }
}

// ...
try {
    let result = await queryKeyBackup(/* ... */);
} catch (e) {
    if (!e isinstance KeyDownloadError) { throw e }
    switch (e.code) { 
       case KeyDownloadErrorCode.VERSION_MISMATCH:
          // etc
    }
}
```

For `RATE_LIMITED`, you might want a second custom error type, which holds the `retryAfterMs` value.

Reply:
If I do that I don't see how to make tests. I can check that the spy has returned with something but I don't see how to check that it has thrown. 
I see `toHaveReturnedWith` methods, but no `toHaveThrown` or `ToHaveRejectWith`. `toThrow` is not working

Reply:
maybe you can do something with [`spiedFunction.mock.results`](https://jestjs.io/docs/mock-function-api#mockfnmockresults)